File: src/l1Metric.js
/**
 * The L1 metric: bisector construction and the upward-test used while walking
 * merge lines. The rest of the pipeline consumes these through a curried
 * `findBisector(siteA, siteB)` function, so this module is the seam where a
 * different metric (e.g. L-infinity) would slot in.
 */

import {samePoint} from './geometry.js';
import {createBisector} from './bisector.js';

/**
 * Curry a bisector factory with fixed canvas dimensions.
 *
 * @param {function} callback
 * @param {number} width
 * @param {number} height
 * @return {function}
 */
export function curryFindBisector(callback, width, height){
    return function(P1, P2){
        return callback(P1, P2, width, height);
    }
}

/**
 * Generate an L1 bisector between two sites.
 *
 * @param {Site} P1
 * @param {Site} P2
 * @param {number} width
 * @param {number} height
 * @returns {Bisector}
 */
export function findL1Bisector(P1, P2, width, height){

    let xDistance = P1.site[0] - P2.site[0];
    let yDistance = P1.site[1] - P2.site[1];

    let midpoint = [
        (P1.site[0] + P2.site[0]) / 2,
        (P1.site[1] + P2.site[1]) / 2
    ];

    let vertexes = [];
    let up = null;

    if(samePoint(P1.site,P2.site)){
        throw new Error(`Duplicate point: Points ${JSON.stringify(P1)} and ${JSON.stringify(P2)} are duplicates. please remove one`);
    }

    if(Math.abs(xDistance) === 0){
        vertexes = [
            [0, midpoint[1]],
            [width, midpoint[1]]
        ];

        return {sites:[P1, P2], up:false, points:vertexes, intersections:[], compound:false};
    }

    if(Math.abs(yDistance) === 0){
        vertexes = [
            [midpoint[0], 0],
            [midpoint[0], height]
        ];

        return {sites:[P1, P2], up:true, points:vertexes, intersections:[], compound:false};
    }

    let slope = yDistance/xDistance > 0 ? -1 : 1;
    let intercept = midpoint[1] - midpoint[0] * slope;

    if(Math.abs(xDistance) > Math.abs(yDistance)){
        vertexes = [
            [(P1.site[1] - intercept) / slope, P1.site[1]],
            [(P2.site[1] - intercept) / slope, P2.site[1]]
        ];

        up = true;
    }
    else if(Math.abs(xDistance) < Math.abs(yDistance)){
        vertexes = [
            [P1.site[0] , (P1.site[0] * slope) + intercept ],
            [P2.site[0] , (P2.site[0] * slope) + intercept ]
        ];

        up = false;
    }
    else { // |dx| === |dy| (square bisector)
        if (slope === 1){
            vertexes = [
                [P1.site[1] - intercept, P1.site[1]],
                [P2.site[1] - intercept, P2.site[1]]
            ];

            up = true;
        }
        else { // slope === -1
            vertexes = [
                [P1.site[0] , -P1.site[0] + intercept ],
                [P2.site[0] , -P2.site[0] + intercept ]
            ];

            up = false;
        }
    }

    let bisector = createBisector([P1, P2], up);

    if(up){
        const sortedVerts = vertexes.sort((a,b) => a[1] - b[1]);

        bisector.points = [
            [sortedVerts[0][0], 0],
            ...sortedVerts,
            [sortedVerts[1][0], height]
        ].sort((a,b) => a[1] - b[1]);

    }
    else{
        const sortedVerts = vertexes.sort((a,b) => a[0] - b[0]);

        bisector.points = [
            [0,sortedVerts[0][1]],
            ...sortedVerts,
            [width,sortedVerts[1][1]]
        ].sort((a,b) => a[0] - b[0]);
    }

    return bisector;
}

/**
 * Check whether the bisector between a hop pair travels upward relative to the
 * merge walk direction. Metric-specific geometry on raw site coordinates.
 *
 * @param {Site} hopTo
 * @param {Site} hopFrom
 * @param {Site} site
 * @param {boolean} goUp
 * @returns {boolean}
 */
export function isNewBisectorUpward(hopTo, hopFrom, site, goUp){

    if(hopTo.site[0] - site.site[0] === 0){
        return site.site[1] > hopTo.site[1];
    }

    let slope = (hopTo.site[1] - site.site[1])/(hopTo.site[0] - site.site[0]);
    let intercept = hopTo.site[1] - (slope * hopTo.site[0]);

    if(Math.abs(slope) === Infinity){
        return site.site[1] > hopTo.site[1];
    }

    let isAboveLine = hopFrom.site[1] > (slope * hopFrom.site[0]) + intercept;

    return isAboveLine;
}

File: py_ai/l1_metric.py
"""
The L1 metric: bisector construction and the upward-test used while walking
merge lines. The rest of the pipeline consumes these through the object
returned by `create_l1_metric`, so this module is the seam where a different
metric (e.g. L-infinity) would slot in.
"""

from geometry import same_point, distance
from bisector import create_bisector


def create_l1_metric(width, height):
    """Build the L1 metric strategy for a fixed canvas."""

    def bisector(p1, p2):
        return find_l1_bisector(p1, p2, width, height)

    return {
        "width": width,
        "height": height,
        "distance": distance,
        "bisector": bisector,
        "is_upward": is_new_bisector_upward,
    }


def find_l1_bisector(p1, p2, width, height):
    """Generate L1 bisector between two sites"""
    x_distance = p1["site"][0] - p2["site"][0]
    y_distance = p1["site"][1] - p2["site"][1]

    mid_x = (p1["site"][0] + p2["site"][0]) / 2
    mid_y = (p1["site"][1] + p2["site"][1]) / 2

    vertexes = []
    up = None

    if same_point(p1["site"], p2["site"]):
        raise ValueError(
            f"Duplicate point: Points {p1['site']} and {p2['site']} are duplicates. Please remove one."
        )

    if abs(x_distance) == 0:
        bisector = create_bisector([p1, p2], False)
        bisector["points"] = [[0, mid_y], [width, mid_y]]
        return bisector

    if abs(y_distance) == 0:
        bisector = create_bisector([p1, p2], True)
        bisector["points"] = [[mid_x, 0], [mid_x, height]]
        return bisector

    slope = -1 if y_distance / x_distance > 0 else 1
    intercept = mid_y - mid_x * slope

    if abs(x_distance) > abs(y_distance):
        v1 = [(p1["site"][1] - intercept) / slope, p1["site"][1]]
        v2 = [(p2["site"][1] - intercept) / slope, p2["site"][1]]
        vertexes = [v1, v2]
        up = True
    elif abs(x_distance) < abs(y_distance):
        v1 = [p1["site"][0], p1["site"][0] * slope + intercept]
        v2 = [p2["site"][0], p2["site"][0] * slope + intercept]
        vertexes = [v1, v2]
        up = False
    else:
        if slope == 1:
            v1 = [p1["site"][1] - intercept, p1["site"][1]]
            v2 = [p2["site"][1] - intercept, p2["site"][1]]
            vertexes = [v1, v2]
            up = True
        else:
            v1 = [p1["site"][0], -p1["site"][0] + intercept]
            v2 = [p2["site"][0], -p2["site"][0] + intercept]
            vertexes = [v1, v2]
            up = False

    bisector = create_bisector([p1, p2], up)

    if up:
        sorted_verts = sorted(vertexes, key=lambda p: p[1])
        bisector["points"] = sorted(
            [[sorted_verts[0][0], 0]] + sorted_verts + [[sorted_verts[1][0], height]],
            key=lambda p: p[1],
        )
    else:
        sorted_verts = sorted(vertexes, key=lambda p: p[0])
        bisector["points"] = sorted(
            [[0, sorted_verts[0][1]]] + sorted_verts + [[width, sorted_verts[1][1]]],
            key=lambda p: p[0],
        )

    return bisector


def is_new_bisector_upward(hop_to, hop_from, site, go_up):
    """Check if bisector is traveling upward or downward"""
    # if hop_to["site"][0] == hop_from["site"][0]:
    #     return site["site"][1] > hop_to["site"][1]

    if hop_to["site"][0] - site["site"][0] == 0:
        return site["site"][1] > hop_to["site"][1]

    slope = (hop_to["site"][1] - site["site"][1]) / (
        hop_to["site"][0] - site["site"][0]
    )
    intercept = hop_to["site"][1] - (slope * hop_to["site"][0])

    is_above_line = hop_from["site"][1] > (slope * hop_from["site"][0]) + intercept

    return is_above_line
